Secret agent hints

balrog-kun opened the pull request:
This is a proposal / RFC for handling the type of agent request made by the IWD wifi backend, where the specific EAP method is not known to NM (IWD doesn't expose that) but specific secrets are being requested using the 'hints' parameter so the client doesn't need to look at the actual connection settings to deduce which secrets might be missing.

Specifically the setting_name parameter in the request is set to "802-1x" and hints is one of "identity", "password" or "private_key_password" in this use case although other values could be supported in the same way.  I'm sort of abusing the EAPMethodSimple class in nm-applet, please see explanation in the commit messages.  It's not very pretty, other ideas to handle this will be welcome

thom311 commented on src/wireless-security/ws-wpa-eap.c:
```diff
@@ -99,7 +99,8 @@ update_secrets (WirelessSecurity *parent, NMConnection *connection)
                 gboolean is_editor,
-                gboolean secrets_only)
+                gboolean secrets_only,
+                const char **secrets_hints)
```
wireless-security is an internal library, both used by libnma and libnm-gtk. You didn't adjust libnm-gtk, so this breaks build `--with-libnm-gtk`.

Maybe use type `const char *const*`?

balrog-kun replied:
Yep, will fix both things.

balrog-kun replied:
Regarding modifying libnm-gtk it seems it is in use by external projects (only gnome-control-center that I could find), would it be better to leave the nma_wifi_dialog_new signature as is and add a new call nma_wifi_dialog_new_for_secrets?

And if this is done in libnm-gtk should the same be done in libnma?

thom311 replied:
libnm-gtk is only on life-support, you don't need add new functionality there at all. Whatever is the least effort. Yes, a new `nma_wifi_dialog_new_for_secrets()` sounds good.

balrog-kun commented:
I have added two patches one of which is slightly unrelated to the rest of the changes.

FYI this 802.1x "hints" functionality got merged into gnome-shell some time ago (granted it's much simpler there).  However it uses gnome-control-center as its connection editor and gnome-control-center uses libnma in some complicated way so it depends on the changes here.

Ping @lkundrak

lkundrak commented:
Hi. Thanks for the patches and your patience. This generally looks okay to me. I'm wondering if you could change `src/libnma/tests/wifi-dialog.c` of `lr/test-programs` branch to demonstrate how is `nma_wifi_dialog_new_for_secrets` used? It would help me test and understand it. A gtkdoc comment would be great too, though I'm quite aware we're missing those in other places.

While reviewing the changes, I've identified a [couple of issues](https://gitlab.gnome.org/GNOME/network-manager-applet/merge_requests/30) -- you may want to include commits from `lr/wifi-dialog-fixes` while testing to avoid the unpleasant surprises I've encountered. (Review of those changes welcome.)

balrog-kun commented:
Hi, yes, with the `lr/wifi-dialog-fixes` fixes things work much better.  I added the gtkdoc comment in this merge request but you might want to improve them.  I also added an example call to the new function in this branch: https://gitlab.gnome.org/balrog-kun/network-manager-applet/commits/lr/test-programs but I needed to rebase it so it's not a clean fork of your test-programs branch.

lkundrak commented:
Thank you, this looks good to me now.

I've merged the branch and it will be included in the next nm-applet release.

bengal commented:
Submitted a follow-up fix here: 

https://gitlab.gnome.org/GNOME/network-manager-applet/merge_requests/39

@balrog-kun can you please review it? Thanks!
